xds/src/main/java/io/grpc/xds/AbstractXdsClient.java

```diff
@@ -0,0 +1,762 @@
+      Node node,
+      ScheduledExecutorService timeService,
+      BackoffPolicy.Provider backoffPolicyProvider,
+      Stopwatch stopwatch) {
+    this.xdsChannel = checkNotNull(channel, "channel");
+    this.node = checkNotNull(node, "node");
+    this.timeService = checkNotNull(timeService, "timeService");
+    this.backoffPolicyProvider = checkNotNull(backoffPolicyProvider, "backoffPolicyProvider");
+    this.stopwatch = checkNotNull(stopwatch, "stopwatch");
+    logId = InternalLogId.allocate("xds-client", null);
+    logger = XdsLogger.withLogId(logId);
+    logger.log(XdsLogLevel.INFO, "Created");
+  }
+
+  /**
+   * Called when an LDS response is received.
+   */
+  // Must be synchronized.
+  protected void handleLdsResponse(String versionInfo, List<Any> resources, String nonce) {
+  }
+
+  /**
+   * Called when a RDS response is received.
+   */
+  // Must be synchronized.
+  protected void handleRdsResponse(String versionInfo, List<Any> resources, String nonce) {
+  }
+
+  /**
+   * Called when a CDS response is received.
+   */
+  // Must be synchronized.
+  protected void handleCdsResponse(String versionInfo, List<Any> resources, String nonce) {
+  }
+
+  /**
+   * Called when an EDS response is received.
+   */
+  // Must be synchronized.
+  protected void handleEdsResponse(String versionInfo, List<Any> resources, String nonce) {
+  }
+
+  /**
+   * Called when the ADS stream is closed passively.
+   */
+  // Must be synchronized.
+  protected void handleStreamClosed(Status error) {
+  }
+
+  /**
+   * Called when the ADS stream has been recreated.
+   */
+  // Must be synchronized.
+  protected void handleStreamRestarted() {
+  }
+
+  /**
+   * Called when being shut down.
+   */
+  // Must be synchronized.
+  protected void handleShutdown() {
+  }
+
+  /**
+   * Synchronizes the execution of the given {@code runnable} with other state mutating operations.
+   */
+  protected abstract void runWithSynchronized(Runnable runnable);
+
+  @Override
+  final void shutdown() {
+    runWithSynchronized(new Runnable() {
+      @Override
+      public void run() {
+        shutdown = true;
+        logger.log(XdsLogLevel.INFO, "Shutting down");
+        xdsChannel.getManagedChannel().shutdown();
+        if (adsStream != null) {
+          adsStream.close(Status.CANCELLED.withDescription("shutdown").asException());
+        }
+        if (rpcRetryTimer != null) {
+          rpcRetryTimer.cancel(false);
+        }
+        handleShutdown();
+      }
+    });
+  }
+
+  @Override
+  public String toString() {
+    return logId.toString();
+  }
+
+  /**
+   * Returns the collection of resources currently subscribing to or {@code null} if not
+   * subscribing to any resources for the given type.
+   *
+   * <p>Note a empty collection indicates subscribing to resources of the given type with
+   * wildcard mode.
+   */
+  // Must be synchronized.
+  @Nullable
+  Collection<String> getSubscribedResources(ResourceType type) {
+    return null;
```

Comment:
This can be made `abstract` instead of returning `null` as I don't see any advantage of providing a default implementation.

Reply:
Sure. Made it abstract.